zerver/worker/queue_processors.py

```diff
@@ -261,9 +263,32 @@ def update_statistics(self, remaining_local_queue_size: int) -> None:
                 f.write(
                     orjson.dumps(stats_dict, option=orjson.OPT_APPEND_NEWLINE | orjson.OPT_INDENT_2)
                 )
+            if backup_old_stats:
+                self.backup_old_stats_if_needed(fn)
             os.rename(tmp_fn, fn)
         self.last_statistics_update_time = time.time()
 
+    def backup_old_stats_if_needed(self, fn: str) -> None:
+        if not os.path.exists(fn):
+            return
+
+        with open(fn, "rb") as f:
+            try:
+                data = orjson.loads(f.read())
+            except orjson.JSONDecodeError:
+                # Backing up an invalid data file is pointless.
+                return
+
+        fn_old = fn + ".old"
+        if data.get("recent_average_consume_time") is None and os.path.exists(fn_old):
+            # If we don't have recent_average_consume_time data,
+            # that means the worker was started recently enough
+            # that it hasn't had time to estimate this number yet.
+            # That makes the it not worth overwriting the old stats file.
+            return
+
+        os.rename(fn, fn_old)
```

Comment:
I think with this algorithm, we should expect there to be a `queue_stats.old` file present at most times, and its value to be very similar to those for the current stats file?  It's probably OK to not try to represent that in the tests, but maybe worth a comment somewhere on that aspect of our plan.

Reply:
Hmm, which values do we expect to be very similar? The .old file will just have the stats from the previous run of the worker, so I'm not sure if we can make any hypothesis on how the .old stats will relate to the new ones :thinking: I added this docstring for now
```
    def backup_old_stats_if_needed(self, fn: str) -> None:
        """
        A backup file with .old suffix will almost always be created upon
        the start of a queue worker. This file will contains the final stats
        from the last *real* run. We emphasize *real*, because runs in which
        no events were processed are discounted.
        """
```